### PR TITLE
event listeners added

### PR DESCRIPTION
**Pull request by kamini-kamal**
The following features have been introduced as a part of this story:

- Event listener for BlockEvent and SmartContractEvent introduced
- The changes have been tested on validator node for smart contract and block. Please find the local snapshots attached:
- smart-contract event
image

- block event

image

### CODE REVIEW

File: src/main/java/hyperledger/besu/java/rest/client/config/EthEventsProperties.java [hunk 1]
```diff
@@ -0,0 +1,18 @@
+package hyperledger.besu.java.rest.client.config;
+
+import java.util.List;
+import lombok.Data;
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.boot.context.properties.ConfigurationProperties;
+import org.springframework.cloud.context.config.annotation.RefreshScope;
+import org.springframework.context.annotation.Configuration;
+
+@Slf4j
```
Comment (nithin-pankaj) [inline src/main/java/hyperledger/besu/java/rest/client/config/EthEventsProperties.java]: Not required.

File: src/main/java/hyperledger/besu/java/rest/client/dto/TopicFilter.java [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+@Data
+@Builder
+public class TopicFilter {
+  private DefaultBlockParameter fromBlock; // optional, params - defaults to latest for both
+  private DefaultBlockParameter toBlock;
+  private String blockHash; // optional, cannot be used together with fromBlock/toBlock
+  private List<String> address; // spec. implies this can be single address as string or list
+  private EthFilter ethFilter;
+
+  @Autowired EthEventsProperties ethEventsProperties;
```
Comment (nithin-pankaj) [inline src/main/java/hyperledger/besu/java/rest/client/dto/TopicFilter.java]: Where do we use `ethEventsProperties` ? Also,  `TopicFilter` is a non Spring managed class, how can we use an Autowired property here?

File: src/main/java/hyperledger/besu/java/rest/client/listener/SmartContractEventListener.java [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+
+@Service
+@Slf4j
+public class SmartContractEventListener implements Runnable {
+  @Autowired private EthConfig ethConfig;
+  @Autowired private TopicFilterHandler topicFilterHandler;
+  @Autowired private EthEventsProperties ethEventsProperties;
+
+  @Override
+  public void run() {
```
Comment (nithin-pankaj) [inline src/main/java/hyperledger/besu/java/rest/client/listener/SmartContractEventListener.java:21]: Who calls this thread?

File: src/main/java/hyperledger/besu/java/rest/client/service/impl/EventPublishServiceImpl.java [hunk 4]
```diff
@@ -72,4 +78,62 @@ public void onFailure(Throwable ex) {
...
+  @Override
+  public boolean publishEventLogs(Object eventLog) {
+    boolean status = true;
+    ProducerRecord<String, String> producerRecord;
+    try {
+
+      Class<?> clazz = eventLog.getClass();
+
+      if (Log.class.equals(clazz)) {
+        assert eventLog instanceof Log;
```
Comment (nithin-pankaj) [inline src/main/java/hyperledger/besu/java/rest/client/service/impl/EventPublishServiceImpl.java]: Why do we need assert? Why not just check `eventLog instanceof Log` in the `if` condition ?

**Review by nithin-pankaj — CHANGES_REQUESTED** [review]

File: src/main/java/hyperledger/besu/java/rest/client/dto/TopicFilter.java [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+import org.web3j.protocol.core.DefaultBlockParameter;
+import org.web3j.protocol.core.methods.request.EthFilter;
+
+@Slf4j
+@Data
+@Builder
+public class TopicFilter {
+  private DefaultBlockParameter fromBlock; // optional, params - defaults to latest for both
+  private DefaultBlockParameter toBlock;
+  private String blockHash; // optional, cannot be used together with fromBlock/toBlock
```
Comment (nidhi-singh02) [inline src/main/java/hyperledger/besu/java/rest/client/dto/TopicFilter.java]: Do we need these parameters in our case ?  Are these getting set anywhere ?

File: src/main/java/hyperledger/besu/java/rest/client/dto/TopicFilter.java [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+    ethFilter = new EthFilter(fromBlock, toBlock, address);
+  }
+
+  private void setEthFilterWithBlockHash() {
+    ethFilter = new EthFilter(blockHash);
+  }
+
+  private void setEthFilterWithBlockHashAndSingleSmartContract(String smartContractAddress) {
+    ethFilter = new EthFilter(blockHash, smartContractAddress);
+  }
```
Comment (nidhi-singh02) [inline src/main/java/hyperledger/besu/java/rest/client/dto/TopicFilter.java]: Are we using it somewhere ? In our case, the application teams won't be passing the start and end block number.

File: src/main/java/hyperledger/besu/java/rest/client/filters/BlockFilterHandler.java [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+  @Autowired(required = false)
+  private EventPublishService eventPublishServiceImpl;
+
+  // To receive all new blocks as they are added to the blockchain (the false parameter specifies
+  // that we only want the blocks, not the embedded transactions too):
+  public void receiveNewlyAddedBlocksOnly() {
+    ethConfig
+        .getWeb3jList()
+        .get(0)
+        .blockFlowable(false)
```
Comment (nidhi-singh02) [inline src/main/java/hyperledger/besu/java/rest/client/filters/BlockFilterHandler.java]: why we changed it to false, earlier it was set as true ?

File: src/main/resources/application.template [hunk 8]
```diff
@@ -40,3 +40,8 @@ kafka:
   event-listener:
     brokerHost: localhost:9093
     topic: besu-offchain-topic
+events:
+  block: true // This is for enabling block events
+  smartcontract:
+  - smartcontractaddress1
+  - smartcontractaddress2
```
Comment (nidhi-singh02) [inline src/main/resources/application.template]: no new line

File: src/main/java/hyperledger/besu/java/rest/client/config/EthEventsProperties.java [hunk 9]
```diff
@@ -0,0 +1,16 @@
...
+import org.springframework.boot.context.properties.ConfigurationProperties;
+import org.springframework.cloud.context.config.annotation.RefreshScope;
+import org.springframework.context.annotation.Configuration;
+
+@Data
+@Configuration
+@ConfigurationProperties(prefix = "events")
+@RefreshScope
+public class EthEventsProperties {
+  private boolean block = false; // initially, setting it to false
```
Comment (abhakish) [inline src/main/java/hyperledger/besu/java/rest/client/config/EthEventsProperties.java]: boolean by default value is false

File: src/main/resources/application-dev.yml [hunk 10]
```diff
@@ -0,0 +1,48 @@
+spring:
```
Comment (abhakish) [inline src/main/resources/application-dev.yml]: This file shouldn't be present, this would be generated when deployment takes place

File: src/main/resources/application.yml [hunk 11]
```diff
@@ -1,5 +1,5 @@
 spring:
```
Comment (abhakish) [inline src/main/resources/application.yml]: Kindly revert the changes, application.template is a template present, don't convert it to application.yml

**Review by abhakish — CHANGES_REQUESTED** [review]

File: src/main/java/hyperledger/besu/java/rest/client/config/EthEventsProperties.java (hunk 1 above)
Comment (nithin-pankaj) [reply src/main/java/hyperledger/besu/java/rest/client/config/EthEventsProperties.java]: Unaddressed.

Comment (kamini-kamal) [reply src/main/java/hyperledger/besu/java/rest/client/config/EthEventsProperties.java]: just addressed!

File: src/main/java/hyperledger/besu/java/rest/client/dto/TopicFilter.java (hunk 2 above)
Comment (kamini-kamal) [reply src/main/java/hyperledger/besu/java/rest/client/dto/TopicFilter.java]: addressed

File: src/main/java/hyperledger/besu/java/rest/client/listener/SmartContractEventListener.java (hunk 3 above)
Comment (kamini-kamal) [reply src/main/java/hyperledger/besu/java/rest/client/listener/SmartContractEventListener.java:21]: implemented the executor service and called the method there

File: src/main/java/hyperledger/besu/java/rest/client/dto/TopicFilter.java (hunk 5 above)
Comment (kamini-kamal) [reply src/main/java/hyperledger/besu/java/rest/client/dto/TopicFilter.java]: addressed

File: src/main/java/hyperledger/besu/java/rest/client/dto/TopicFilter.java (hunk 6 above)
Comment (kamini-kamal) [reply src/main/java/hyperledger/besu/java/rest/client/dto/TopicFilter.java]: corrected

File: src/main/java/hyperledger/besu/java/rest/client/filters/BlockFilterHandler.java (hunk 7 above)
Comment (kamini-kamal) [reply src/main/java/hyperledger/besu/java/rest/client/filters/BlockFilterHandler.java]: corrected!

File: src/main/java/hyperledger/besu/java/rest/client/config/EthEventsProperties.java (hunk 9 above)
Comment (kamini-kamal) [reply src/main/java/hyperledger/besu/java/rest/client/config/EthEventsProperties.java]: corrected

File: src/main/resources/application-dev.yml (hunk 10 above)
Comment (kamini-kamal) [reply src/main/resources/application-dev.yml]: removed

File: src/main/resources/application.yml (hunk 11 above)
Comment (kamini-kamal) [reply src/main/resources/application.yml]: removed

File: src/main/resources/application.template (hunk 8 above)
Comment (kamini-kamal) [reply src/main/resources/application.template]: corrected

File: src/main/java/hyperledger/besu/java/rest/client/filters/BlockFilterHandler.java [hunk 12]
```diff
@@ -0,0 +1,41 @@
...
+        .blockFlowable(true)
+        .doOnError(
+            error ->
+                log.error(
+                    "Error occurred while listening to receiveNewlyAddedBlocksOnly events"
+                        + error.getMessage()))
+        .subscribe(
+            block -> {
+              if (!block.hasError()) {
+                log.info("block event listening {}", block.getBlock());
```
Comment (nidhi-singh02) [inline src/main/java/hyperledger/besu/java/rest/client/filters/BlockFilterHandler.java]: I don't see the block data in the attached reference - it's EthBlock object, Can we check what is being published to kafka.

File: src/main/java/hyperledger/besu/java/rest/client/config/EthConfig.java [hunk 13]
```diff
@@ -26,7 +27,7 @@ public class EthConfig {
 
   @RefreshScope
   @Bean
-  public Credentials getCredentials() {
+  public CredentialWrapper getCredentials() {
```
Comment (nidhi-singh02) [inline src/main/java/hyperledger/besu/java/rest/client/config/EthConfig.java:30]: Q: Why we introduced `CredentialWrapper` ?

File: src/main/java/hyperledger/besu/java/rest/client/config/TaskExecutorConfig.java [hunk 14]
```diff
@@ -0,0 +1,21 @@
...
+
+@Configuration
+public class TaskExecutorConfig {
+
+  @Bean
+  public TaskExecutor threadPoolTaskExecutor() {
+
+    ThreadPoolTaskExecutor threadPoolTaskExecutor = new ThreadPoolTaskExecutor();
+    threadPoolTaskExecutor.setCorePoolSize(7);
+    threadPoolTaskExecutor.setMaxPoolSize(7);
```
Comment (nidhi-singh02) [inline src/main/java/hyperledger/besu/java/rest/client/config/TaskExecutorConfig.java]: Let's make these parameters as configurable, let's not hardcode it. We can have these as default values.

File: src/main/java/hyperledger/besu/java/rest/client/filters/TransactionFilterHandler.java [hunk 15]
```diff
@@ -0,0 +1,60 @@
...
+        .doOnError(
+            error ->
+                log.error(
+                    "Error occurred while listening to receiveAllPendingTransactions events"
+                        + error.getMessage()))
+        .subscribe(tx -> {});
+  }
+
+  // Replay all blocks to the most current, but with transactions contained within blocks:
+  private void replayPastAndFutureTransactions(DefaultBlockParameter startBlock) {
```
Comment (nidhi-singh02) [inline src/main/java/hyperledger/besu/java/rest/client/filters/TransactionFilterHandler.java]: Q: When this function will be called ? Are we invoking it in any case.

File: src/main/java/hyperledger/besu/java/rest/client/config/EthConfig.java (hunk 13 above)
Comment (kamini-kamal) [reply src/main/java/hyperledger/besu/java/rest/client/config/EthConfig.java:30]: CGLIB proxy cannot create beans for private constructor, so created a wrapper. This is needed when using @RefreshScope annotation

File: src/main/java/hyperledger/besu/java/rest/client/filters/BlockFilterHandler.java (hunk 12 above)
Comment (kamini-kamal) [reply src/main/java/hyperledger/besu/java/rest/client/filters/BlockFilterHandler.java]: Extracting the following values before sending to kafka for block events:

- ETH_BLOCK_HASH
- EVENT_TYPE
- eventLog which consists of EthBlock details like transactions, timestamp, gasUsed, gasLimit, transactionsRoot, number, hash etc.

File: src/main/java/hyperledger/besu/java/rest/client/filters/TransactionFilterHandler.java (hunk 15 above)
Comment (kamini-kamal) [reply src/main/java/hyperledger/besu/java/rest/client/filters/TransactionFilterHandler.java]: Currently, we are not invoking it. I marked it for future scope . Let me know if that's ok

File: src/main/java/hyperledger/besu/java/rest/client/exception/ErrorCode.java [hunk 16]
```diff
@@ -12,6 +12,7 @@ public enum ErrorCode {
   NOT_SUPPORTED(1002, "Operation passed not supported."),
   VALIDATION_FAILED(1003, "The data passed fails validation from the resource."),
   INITIALIZATION_FAILED(1004, "Failed to bootstrap the applciation"),
+  OPERATION_NOT_SUPPORTED(1005, "Operation not supported"),
```
Comment (abhakish) [inline src/main/java/hyperledger/besu/java/rest/client/exception/ErrorCode.java]: `NOT_SUPPORTED(1002, "Operation passed not supported."),` already present

File: src/main/java/hyperledger/besu/java/rest/client/exception/ErrorCode.java [hunk 17]
```diff
@@ -33,7 +34,9 @@ public enum ErrorCode {
 
   NOT_DEFINED(
       9999,
-      "The exception is not a BaseException OR error code is not yet defined by the developer");
+      "The exception is not a BaseException OR error code is not yet defined by the developer"),
+
+  INVALID_ARGUMENT_FOUND(1006, "Invalid Argument Found");
```
Comment (abhakish) [inline src/main/java/hyperledger/besu/java/rest/client/exception/ErrorCode.java]: [nitpick] place this with rest of the `1XXX` error codes

File: src/main/java/hyperledger/besu/java/rest/client/filters/TopicFilterHandler.java [hunk 18]
```diff
@@ -0,0 +1,70 @@
...
+                log.info("ex {}", e.getMessage());
+              }
+            });
+  }
+
+  private void createAndSendSmartContractEvents(Log eventLog) {
+    eventLog.getTopics().stream()
+        .forEach(
+            topic -> {
+              String eventName =
```
Comment (abhakish) [inline src/main/java/hyperledger/besu/java/rest/client/filters/TopicFilterHandler.java]: extract a method for this operation

File: src/main/java/hyperledger/besu/java/rest/client/filters/TopicFilterHandler.java [hunk 19]
```diff
@@ -0,0 +1,70 @@
...
+                smartContractEventWrapper =
+                    SmartContractEventWrapper.builder()
+                        .eventName(eventName)
+                        .data(SmartContractUtility.convertToStringFromHex(eventLog.getData()))
+                        .log(eventLog)
+                        .build();
+              } catch (DecoderException e) {
+                throw new RuntimeException(e);
+              } catch (UnsupportedEncodingException e) {
+                throw new RuntimeException(e);
```
Comment (abhakish) [inline src/main/java/hyperledger/besu/java/rest/client/filters/TopicFilterHandler.java]: throw custom exceptions rather than generic ones

File: src/main/java/hyperledger/besu/java/rest/client/listener/BlockEventListener.java [hunk 20]
```diff
@@ -1,30 +1,24 @@
...
 @Service
+@ConditionalOnProperty(value = "events.block", havingValue = "true")
 public class BlockEventListener implements Runnable {
 
   @Autowired private EthConfig ethConfig;
+  @Autowired private BlockFilterHandler blockFilterHandler;
 
   @Override
   public void run() {
     // TODO: Improve the logic to include all nodes from the network
```
Comment (abhakish) [inline src/main/java/hyperledger/besu/java/rest/client/listener/BlockEventListener.java:20]: [nitpick] is this TODO still required

File: src/main/java/hyperledger/besu/java/rest/client/utils/SmartContractUtility.java [hunk 21]
```diff
@@ -0,0 +1,422 @@
...
+
+  public Map<String, String> generateEventHashFromAbiFiles(
+      EthEventsProperties.SmartContract smartContract) {
+    Map<String, String> eventNameByHash = new HashMap<>();
+    List<String> abiFileList = getAbiFilePathListForSmartContract(smartContract);
+    abiFileList.stream().forEach(abiFilePath -> addEventNameByHash(abiFilePath, eventNameByHash));
+    return eventNameByHash;
+  }
+
+  private static TypeReference<? extends Type> getType(String type, boolean indexed) {
```
Comment (abhakish) [inline src/main/java/hyperledger/besu/java/rest/client/utils/SmartContractUtility.java]: instead of defining this switch cases again in this method, you can do 
```
Class clazz = AbiTypes.getType(type);
return TypeReference.create(clazz,indexed);
```

File: src/main/java/hyperledger/besu/java/rest/client/utils/SmartContractUtility.java [hunk 22]
```diff
@@ -0,0 +1,422 @@
...
+    try {
+      mapper
+          .readValue(
+              abiString,
+              new com.fasterxml.jackson.core.type.TypeReference<List<AbiDefinition>>() {})
+          .stream()
+          .forEach(abiDef -> createEventDefinition(abiDef, eventNameByHash));
+
+    } catch (JsonProcessingException e) {
+      throw new RuntimeException(e);
```
Comment (abhakish) [inline src/main/java/hyperledger/besu/java/rest/client/utils/SmartContractUtility.java]: throw custom exceptions

**Review by abhakish — CHANGES_REQUESTED** [review]

File: src/main/java/hyperledger/besu/java/rest/client/service/impl/EventPublishServiceImpl.java (hunk 4 above)
Comment (kamini-kamal) [reply src/main/java/hyperledger/besu/java/rest/client/service/impl/EventPublishServiceImpl.java]: corrected

File: src/main/java/hyperledger/besu/java/rest/client/config/TaskExecutorConfig.java (hunk 14 above)
Comment (kamini-kamal) [reply src/main/java/hyperledger/besu/java/rest/client/config/TaskExecutorConfig.java]: incorporated the changes

File: src/main/java/hyperledger/besu/java/rest/client/listener/BlockEventListener.java (hunk 20 above)
Comment (kamini-kamal) [reply src/main/java/hyperledger/besu/java/rest/client/listener/BlockEventListener.java:20]: Yes.

File: src/main/java/hyperledger/besu/java/rest/client/utils/SmartContractUtility.java (hunk 21 above)
Comment (kamini-kamal) [reply src/main/java/hyperledger/besu/java/rest/client/utils/SmartContractUtility.java]: addressed. Thanks for the suggestion

File: src/main/java/hyperledger/besu/java/rest/client/EthConnectorApplication.java [hunk 23]
```diff
@@ -6,11 +6,13 @@
 import org.springframework.boot.builder.SpringApplicationBuilder;
 import org.springframework.boot.context.properties.EnableConfigurationProperties;
 import org.springframework.boot.web.servlet.support.SpringBootServletInitializer;
+import org.springframework.scheduling.annotation.EnableAsync;
 
 /** Bootstrap for the Eth-Connector application. */
 @Slf4j
 @SpringBootApplication
 @EnableConfigurationProperties
+@EnableAsync
```
Comment (nidhi-singh02) [inline src/main/java/hyperledger/besu/java/rest/client/EthConnectorApplication.java:15]: Why we added this annonation ?

File: src/main/java/hyperledger/besu/java/rest/client/utils/SmartContractUtility.java [hunk 24]
```diff
@@ -0,0 +1,101 @@
...
+    List<String> abiFileList = new ArrayList<>();
+    String baseAbiPath = smartContract.getAbiBasePath();
+    smartContract.getAddresses().stream()
+        .forEach(address -> getAbiList(baseAbiPath, address, abiFileList));
+    return abiFileList;
+  }
+
+  private void getAbiList(String baseAbiPath, String parentFolderName, List<String> abiFileList) {
+    String folderPath = baseAbiPath + parentFolderName;
+    File[] files = new File(folderPath).listFiles();
```
Comment (nidhi-singh02) [inline src/main/java/hyperledger/besu/java/rest/client/utils/SmartContractUtility.java]: Let's give meaningful names. These names are too generic.

File: src/main/java/hyperledger/besu/java/rest/client/service/impl/EventPublishServiceImpl.java [hunk 25]
```diff
@@ -72,4 +79,68 @@ public void onFailure(Throwable ex) {
...
+
+      producerRecord.headers().add(new RecordHeader(EVENT_TYPE, EVENT_TYPE_LOG.getBytes()));
+      ListenableFuture<SendResult<String, String>> future = kafkaTemplate.send(producerRecord);
+
+      future.addCallback(
+          new ListenableFutureCallback<SendResult<String, String>>() {
+
+            @Override
+            public void onSuccess(SendResult<String, String> result) {
+              log.info(
```
Comment (nidhi-singh02) [inline src/main/java/hyperledger/besu/java/rest/client/service/impl/EventPublishServiceImpl.java]: nitpick log.debug

File: src/main/java/hyperledger/besu/java/rest/client/EthConnectorApplication.java (hunk 23 above)
Comment (kamini-kamal) [reply src/main/java/hyperledger/besu/java/rest/client/EthConnectorApplication.java:15]: switches on Spring's ability to run @Async methods in a background thread pool

File: src/main/java/hyperledger/besu/java/rest/client/utils/SmartContractUtility.java (hunk 24 above)
Comment (kamini-kamal) [reply src/main/java/hyperledger/besu/java/rest/client/utils/SmartContractUtility.java]: addressed

File: src/main/java/hyperledger/besu/java/rest/client/service/impl/EventPublishServiceImpl.java (hunk 25 above)
Comment (kamini-kamal) [reply src/main/java/hyperledger/besu/java/rest/client/service/impl/EventPublishServiceImpl.java]: addressed

File: src/main/java/hyperledger/besu/java/rest/client/config/EthConfig.java (hunk 13 above)
Comment (nidhi-singh02) [reply src/main/java/hyperledger/besu/java/rest/client/config/EthConfig.java:30]: @arsulegai refer here, this is the fix I was pointing you to.

Comment (arsulegai) [reply src/main/java/hyperledger/besu/java/rest/client/config/EthConfig.java:30]: Thanks, nice catch @kamini-kamal

**Review by arsulegai — APPROVED** [review]

File: src/main/java/hyperledger/besu/java/rest/client/exception/ErrorCode.java [hunk 26]
```diff
@@ -11,7 +11,9 @@ public enum ErrorCode {
   NO_EVENTS_FOUND(1001, "Not able to find any events for the passed input."),
   NOT_SUPPORTED(1002, "Operation passed not supported."),
   VALIDATION_FAILED(1003, "The data passed fails validation from the resource."),
-  INITIALIZATION_FAILED(1004, "Failed to bootstrap the application"),
+  INITIALIZATION_FAILED(1004, "Failed to bootstrap the applciation"),
```
Comment (nidhi-singh02) [inline src/main/java/hyperledger/besu/java/rest/client/exception/ErrorCode.java]: [nitpick] typo : application

File: src/main/java/hyperledger/besu/java/rest/client/filters/TransactionFilterHandler.java (hunk 15 above)
Comment (nidhi-singh02) [reply src/main/java/hyperledger/besu/java/rest/client/filters/TransactionFilterHandler.java]: Can we add a comment for the same.

File: src/main/java/hyperledger/besu/java/rest/client/utils/SmartContractUtility.java (hunk 24 above)
Comment (nidhi-singh02) [reply src/main/java/hyperledger/besu/java/rest/client/utils/SmartContractUtility.java]: I still see the names "files" and "folderPath"

File: src/main/java/hyperledger/besu/java/rest/client/service/impl/TransactionServiceImpl.java [hunk 27]
```diff
@@ -41,22 +41,25 @@
 @Service
 public class TransactionServiceImpl implements TransactionService {
 
+  @Autowired ObjectMapper objectMapper;
```
Comment (nithin-pankaj) [inline src/main/java/hyperledger/besu/java/rest/client/service/impl/TransactionServiceImpl.java]: private

File: src/main/java/hyperledger/besu/java/rest/client/service/impl/TopicFilterServiceImpl.java [hunk 28]
```diff
@@ -0,0 +1,22 @@
...
+
+@Service
+// This service is used for Topic filter related tasks
+public class TopicFilterServiceImpl implements TopicFilterService {
+  private static final DefaultBlockParameter fromBlock =
+      DefaultBlockParameterName.EARLIEST; // optional, params - defaults to latest for both
+  private static final DefaultBlockParameter toBlock = DefaultBlockParameterName.LATEST;
+  @Autowired private EthEventsProperties ethEventsProperties;
+
+  public EthFilter createEthFilterWithSmartContractAddressList() {
```
Comment (nithin-pankaj) [inline src/main/java/hyperledger/besu/java/rest/client/service/impl/TopicFilterServiceImpl.java:20]: `@Override` ?

File: src/main/java/hyperledger/besu/java/rest/client/config/EthEventsProperties.java [hunk 29]
```diff
@@ -0,0 +1,45 @@
...
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.boot.context.properties.ConfigurationProperties;
+import org.springframework.cloud.context.config.annotation.RefreshScope;
+import org.springframework.context.annotation.Configuration;
+
+@Data
+@Configuration
+@ConfigurationProperties(prefix = "events")
+@RefreshScope
+@Slf4j
```
Comment (nithin-pankaj) [inline src/main/java/hyperledger/besu/java/rest/client/config/EthEventsProperties.java]: remove if nothing is being logged in this Class.

File: src/main/java/hyperledger/besu/java/rest/client/filters/BlockFilterHandler.java [hunk 30]
```diff
@@ -0,0 +1,41 @@
...
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.scheduling.annotation.Async;
+import org.springframework.stereotype.Service;
+
+@Service
+@Slf4j
+public class BlockFilterHandler {
+  @Autowired private EthConfig ethConfig;
+
+  @Autowired(required = false)
```
Comment (nithin-pankaj) [inline src/main/java/hyperledger/besu/java/rest/client/filters/BlockFilterHandler.java]: Why optional?
The only method in this service has a strict dependency with this bean.

File: src/main/java/hyperledger/besu/java/rest/client/filters/TopicFilterHandler.java [hunk 31]
```diff
@@ -0,0 +1,77 @@
...
+import org.springframework.scheduling.annotation.Async;
+import org.springframework.stereotype.Service;
+import org.web3j.protocol.core.methods.response.Log;
+
+@Service
+@Slf4j
+public class TopicFilterHandler {
+  @Autowired private EthConfig ethConfig;
+
+  @Autowired(required = false)
```
Comment (nithin-pankaj) [inline src/main/java/hyperledger/besu/java/rest/client/filters/TopicFilterHandler.java]: Again, why optional if the methods of this class depends on this?

File: src/main/java/hyperledger/besu/java/rest/client/filters/TransactionFilterHandler.java [hunk 32]
```diff
@@ -0,0 +1,60 @@
...
+import hyperledger.besu.java.rest.client.config.EthConfig;
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Service;
+import org.web3j.protocol.core.DefaultBlockParameter;
+
+@Service
+@Slf4j
+// This class is for the future scope, this is currently not attached to the event listeners
+public class TransactionFilterHandler {
```
Comment (nithin-pankaj) [inline src/main/java/hyperledger/besu/java/rest/client/filters/TransactionFilterHandler.java:13]: What's the purpose of creating this Service bean with all private methods?? Was this a requirement?

**Review by nithin-pankaj — CHANGES_REQUESTED** [review]

File: src/main/java/hyperledger/besu/java/rest/client/filters/TransactionFilterHandler.java (hunk 15 above)
Comment (kamini-kamal) [reply src/main/java/hyperledger/besu/java/rest/client/filters/TransactionFilterHandler.java]: yes added

File: src/main/java/hyperledger/besu/java/rest/client/utils/SmartContractUtility.java (hunk 24 above)
Comment (kamini-kamal) [reply src/main/java/hyperledger/besu/java/rest/client/utils/SmartContractUtility.java]: addressed

File: src/main/java/hyperledger/besu/java/rest/client/exception/ErrorCode.java (hunk 26 above)
Comment (kamini-kamal) [reply src/main/java/hyperledger/besu/java/rest/client/exception/ErrorCode.java]: addressed

File: src/main/java/hyperledger/besu/java/rest/client/service/impl/TopicFilterServiceImpl.java (hunk 28 above)
Comment (kamini-kamal) [reply src/main/java/hyperledger/besu/java/rest/client/service/impl/TopicFilterServiceImpl.java:20]: addressed

File: src/main/java/hyperledger/besu/java/rest/client/service/impl/TransactionServiceImpl.java (hunk 27 above)
Comment (kamini-kamal) [reply src/main/java/hyperledger/besu/java/rest/client/service/impl/TransactionServiceImpl.java]: addressed

File: src/main/java/hyperledger/besu/java/rest/client/config/EthEventsProperties.java (hunk 29 above)
Comment (kamini-kamal) [reply src/main/java/hyperledger/besu/java/rest/client/config/EthEventsProperties.java]: addressed

File: src/main/java/hyperledger/besu/java/rest/client/filters/BlockFilterHandler.java (hunk 30 above)
Comment (kamini-kamal) [reply src/main/java/hyperledger/besu/java/rest/client/filters/BlockFilterHandler.java]: addressed

File: src/main/java/hyperledger/besu/java/rest/client/filters/TopicFilterHandler.java (hunk 31 above)
Comment (kamini-kamal) [reply src/main/java/hyperledger/besu/java/rest/client/filters/TopicFilterHandler.java]: addressed

File: src/main/java/hyperledger/besu/java/rest/client/filters/TransactionFilterHandler.java (hunk 32 above)
Comment (kamini-kamal) [reply src/main/java/hyperledger/besu/java/rest/client/filters/TransactionFilterHandler.java:13]: I made them public and sync

**Review by nidhi-singh02 — APPROVED** [review]

File: src/main/java/hyperledger/besu/java/rest/client/listener/SmartContractEventListener.java [hunk 33]
```diff
@@ -0,0 +1,29 @@
...
+import hyperledger.besu.java.rest.client.exception.ErrorCode;
+import hyperledger.besu.java.rest.client.exception.NotFoundException;
+import hyperledger.besu.java.rest.client.filters.TopicFilterHandler;
+import org.apache.commons.lang3.ObjectUtils;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.context.annotation.Scope;
+import org.springframework.stereotype.Component;
+
+@Component
+@Scope("prototype")
```
Comment (nithin-pankaj) [inline src/main/java/hyperledger/besu/java/rest/client/listener/SmartContractEventListener.java:14]: Why prototype scoped? Why wouldn't default scope work here?

File: src/main/java/hyperledger/besu/java/rest/client/service/AsyncEventListener.java [hunk 34]
```diff
@@ -0,0 +1,39 @@
...
+  @Autowired private ApplicationContext appContext;
+
+  @Autowired private TaskExecutor taskExecutor;
+  @Autowired private EthEventsProperties ethEventsProperties;
+
+  @EventListener(ApplicationReadyEvent.class)
+  public void smartContractEventListenerThread() {
+    SmartContractEventListener smartContractEventListener =
+        appContext.getBean(SmartContractEventListener.class);
+    taskExecutor.execute(smartContractEventListener);
```
Comment (nithin-pankaj) [inline src/main/java/hyperledger/besu/java/rest/client/service/AsyncEventListener.java:27]: This task shouldn't be executed if address list is empty, right? I can see a similar check in the adjascent method `blockEventListenerThread`

P.S rename these methods to something meaningful like 'runSmartContractEventListening'.

File: src/main/java/hyperledger/besu/java/rest/client/service/impl/EventPublishServiceImpl.java [hunk 35]
```diff
@@ -72,4 +79,68 @@ public void onFailure(Throwable ex) {
...
+                new RecordHeader(
+                    CONTRACT_ADDRESS,
+                    ((SmartContractEventWrapper) eventLog).getLog().getAddress().getBytes()));
+        producerRecord
+            .headers()
+            .add(
+                new RecordHeader(
+                    EVENT_NAME, ((SmartContractEventWrapper) eventLog).getEventName().getBytes()));
+      } else if (eventLog instanceof EthBlock) {
+        producerRecord = new ProducerRecord<>(topicName, mapper.writeValueAsString(eventLog));
```
Comment (nithin-pankaj) [inline src/main/java/hyperledger/besu/java/rest/client/service/impl/EventPublishServiceImpl.java:106]: casting?

File: src/main/java/hyperledger/besu/java/rest/client/utils/SmartContractUtility.java [hunk 36]
```diff
@@ -0,0 +1,107 @@
...
+    String baseAbiPath = smartContract.getAbiBasePath();
+    smartContract.getAddresses().stream()
+        .forEach(address -> getAbiListForSmartContracts(baseAbiPath, address, abiFileList));
+    return abiFileList;
+  }
+
+  private void getAbiListForSmartContracts(
+      String baseAbiPath, String parentFolderName, List<String> abiFileList) {
+    String smartContractPath = baseAbiPath + parentFolderName;
+    File[] abiFiles = new File(smartContractPath).listFiles();
```
Comment (nithin-pankaj) [inline src/main/java/hyperledger/besu/java/rest/client/utils/SmartContractUtility.java:68]: There needs to be a validation some place to check if file exists here. Otherwise this would probably throw an NPE.

**Review by nithin-pankaj — CHANGES_REQUESTED** [review]
